metricflow/collection_helpers/pretty_print.py

```diff
@@ -0,0 +1,451 @@
+from __future__ import annotations
+
+import logging
+import pprint
+from collections.abc import Mapping
+from dataclasses import fields, is_dataclass
+from enum import Enum
+from typing import Any, Dict, List, Optional, Sized, Union
+
+from pydantic import BaseModel
+
+from metricflow.formatting import indent_log_line
+
+logger = logging.getLogger(__name__)
+
+
+class MetricFlowPrettyFormatter:
+    """Creates string representations of objects useful for logging / snapshots."""
+
+    def __init__(
+        self,
+        indent_prefix: str,
+        max_line_length: int,
+        include_object_field_names: bool,
+        include_none_object_fields: bool,
+        include_empty_object_fields: bool,
+    ) -> None:
+        """See mf_pformat() for argument descriptions."""
+        self._indent_prefix = indent_prefix
+        if max_line_length <= 0:
+            raise ValueError(f"max_line_length must be > 0 as required by pprint.pformat(). Got {max_line_length}")
+        self._max_line_width = max_line_length
+        self._include_object_field_names = include_object_field_names
+        self._include_none_object_fields = include_none_object_fields
+        self._include_empty_object_fields = include_empty_object_fields
+
+    @staticmethod
+    def _is_pydantic_base_model(obj: Any):  # type:ignore
+        return isinstance(obj, BaseModel)
+
+    def _handle_sequence_obj(self, list_like_obj: Union[list, tuple], remaining_line_width: Optional[int]) -> str:
+        """Pretty prints a sequence object i.e. list or tuple.
+
+        Args:
+            list_like_obj: A list or a tuple.
+            remaining_line_width: If specified, try to make the string representation <= this many columns wide.
+
+        Returns:
+            A string representation of the sequence like (1,) or [1, 2].
+        """
+        if isinstance(list_like_obj, list):
+            left_enclose_str = "["
+            right_enclose_str = "]"
+        elif isinstance(list_like_obj, tuple):
+            left_enclose_str = "("
+            right_enclose_str = ")"
+        else:
+            raise RuntimeError(f"Unhandled type: {type(list_like_obj)}")
```

Comment:
This should be unreachable although it's fine to have it in case someone decides to override the typechecker.

Relatedly, sets are a thing, so we might consider adding support for those. They're represented as `"{1, 2, 3}"`